### ISSUE TITLE
Bug: `getNodeByRangeIndex` returns incorrect node for decorated `ClassDeclaration` inside `ExportDeclaration`

### ISSUE THREAD

**Issue by 9romise** [typescript-eslint/typescript-eslint#11578]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&showAST=es&fileType=.ts&code=AIMw9mAUDeBQAE8B2BDAtgUwFzwIwBpYBfASlgwA8AHMAJwBd4BjAGxQGd34AxCeaeEVhA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

Please focus on the range of the nodes.

### Repro Code

```TypeScript
@foo({
  name:    1,
})
export class Foo { }
```

### Expected Result

- Run `sourceCode.getNodeByRangeIndex(14)`
- Return `Property`

### Actual Result

Return `Program`

### Additional Info

When a `ClassDeclaration` has a parent node of type `ExportNamedDeclaration` or `ExportDefaultDeclaration`, `sourceCode.getNodeByRangeIndex` fails to return the correct node based on the provided index.

This occurs because the range of `ExportDeclaration` and `ClassDeclaration` nodes does not include the range of decorators, causing decorators to be skipped during node retrieval.

This appears to be a design limitation in the ESTree specification. However, is it possible to override the `getNodeByRangeIndex` to ensure it returns the correct node in such cases?

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.41.0` |
| `@typescript-eslint/parser`            | `8.41.0` |
| `@typescript-eslint/rule-tester`       | `8.41.0` |
| `@typescript-eslint/scope-manager`     | `8.41.0` |
| `@typescript-eslint/typescript-estree` | `8.41.0` |
| `@typescript-eslint/type-utils`        | `8.41.0` |
| `@typescript-eslint/utils`             | `8.41.0` |
| `TypeScript`                           | `5.8.3` |
| `ESLint`                               | `9.34.0` |
| `node`                                 | `22.19.0` |

**Comment by JoshuaKGoldberg** [comment]
Fun! Yet another spooky behavior with decorators. Looking at the ESTree view in the playground, it indeed shows:

* Program: [0, 42]
  * ExportNamedDeclaration: [21, 41]
    * ClassDeclaration: [28, 41]
      * Decorator: [0, 20]

I don't know what is expected here. The ClassDeclaration itself doesn't extend earlier than character 28, but I'm not sure whether the ExportNamedDeclaration should. @bradzacher @fisker?

**Comment by 9romise** [comment]
I think that with the current AST design, it's impossible for the node's range to be correctly described in this case.

What this issue expected is that, when writing a custom rule, calling `sourceCode.getNodeByRangeIndex` for such nodes can correctly return the expected node. If the description is hard to follow and you need a minimal reproduction, just let me know, I'm happy to provide one.

**Comment by fisker** [comment]
Personally, I prefer `Class{Declaration,Expression}` and `ExportNamedDeclaration` range to include `decorators`.

There was a discussion on ESTree https://github.com/estree/estree/issues/315.
The range issue also raised on Meriyah https://github.com/meriyah/meriyah/issues/421

Currently, Prettier extends the range of nodes above to include `decorators`. https://github.com/prettier/prettier/blob/267518dc49e5f639bfd534253f272fbe105a3b68/src/language-js/loc.js#L16-L19

**Comment by bradzacher** [comment]
It's a bit messy as there's ultimately no right answer here based on existing AST behaviours and things get messy no matter which way you go.

I get why they allowed decorators there but it's also such a cooked thing lol.

Does espree support them? What did they decide on? I know that we never fully came to a decision in the estree issue.

**Comment by JoshuaKGoldberg** [comment]
@bradzacher @fisker based on https://github.com/estree/estree/issues/315#issuecomment-2032405347 it sounds like we're leaning towards the Babel approach of children nodes always being within their parents' ranges? As in, accepting this change to have the start of a decorated class go earlier to include its earliest decorator?

**Comment by fisker** [comment]
Ready to break `sourceCode.getFirstToken()`!

**Comment by bradzacher** [comment]
How does upstream work here? What does espree do?

**Comment by fisker** [comment]
Why do you think Espree supports it? It's still at stage 3.

**Comment by bradzacher** [comment]
They support stage 3 syntax. But it looks like acorn hasn't added support for it yet.

**Comment by JoshuaKGoldberg** [comment]
Marking as accepting PRs, with the note that it's a breaking change that will have to wait until our next major version. Which has no estimated date and will probably not be for a very long time.
